scripts/spec_validator.py

```diff
@@ -0,0 +1,123 @@
+    with open(path, "r") as file:
+        spec = yaml.safe_load(file)
+        definitions = spec["definitions"]
+        return flatten(definitions)
+
+
+def filter_files(previous, current):
+    # Adding new packages is ok so we only want to do a 1:1
+    # comparison between the new files and the old ones
+    files = [os.path.basename(path) for path in previous]
+    return [path for path in current if os.path.basename(path) in files]
+
+
+def validate_bitfields(previous_flag, current_flag):
+    for previous, current in zip(previous_flag["fields"], current_flag["fields"]):
+        previous_bitfield = unpack(previous)[0]
+        current_bitfield = unpack(current)[0]
+        if previous_bitfield != current_bitfield:
+            raise RuntimeError(
+                "Breaking Message Mutation Detected!\n"
+                f"Previous: {previous_bitfield}, Current: {current_bitfield}"
+            )
+
+
+def validate_fields(previous_fields, current_fields):
+    for previous, current in zip(previous_fields, current_fields):
+        previous_key, previous_value = unpack(previous)
+        current_key, current_value = unpack(current)
+        if "fields" in previous_value:
+            logging.info(f"Processing Bitfield: {previous_key}")
+            validate_bitfields(previous_value, current_value)
+        previous_type = previous_value["type"]
+        current_type = current_value["type"]
+        if current_type != previous_type:
+            raise RuntimeError(
+                "Breaking Message Mutation Detected!\n"
+                f"Field (previous, current): {previous_key}, {current_key}\n"
+                f"Type (previous, current): {previous_type}, {current_type}"
+            )
+
+
+def validate_id(previous, current):
+    prev_id = previous["id"]
+    curr_id = current["id"]
+    if prev_id != curr_id:
+        logging.info("Message ID Change Detected!")
+        logging.info(f"Previous ID: {prev_id}, Current ID: {curr_id}")
+        raise RuntimeError(
+            "Message ID Change Detected!\n"
+            f"Previous ID: {prev_id}, Current ID: {curr_id}"
+        )
+
+
+def validate_definitions(previous_definitions, current_definitions):
+    for key in previous_definitions.keys():
+        logging.info(f"Processing Definition: {key}")
+        previous = previous_definitions[key]
+        current = current_definitions[key]
+        if is_message(key):
+            validate_id(previous, current)
+        # bail if message has no fields
+        # assumes adding fields to an empty message is ok.
+        if "fields" not in previous:
+            continue
+        previous_fields = previous["fields"]
+        current_fields = current["fields"]
+        validate_fields(previous_fields, current_fields)
+
+
+def validate(previous, current):
+    previous_files = glob.glob(f"{previous}/*.yaml")
+    current_files = glob.glob(f"{current}/*.yaml")
+    current_files = filter_files(previous_files, current_files)
```

Comment:
Not sure we need to handle it in this PR, but it might be worth detecting the (degenerate) scenario where a package was renamed and asserting if that happens.